nautobot_circuit_maintenance/tests/test_sources.py

```diff
@@ -0,0 +1,177 @@
+"""Test sourcess utils."""
+from unittest.mock import Mock, patch
+from django.conf import settings
+from django.test import TestCase
+from parameterized import parameterized
+from pydantic.error_wrappers import ValidationError  # pylint: disable=no-name-in-module
+from nautobot.circuits.models import Provider
+
+from nautobot_circuit_maintenance.models import NotificationSource
+
+from nautobot_circuit_maintenance.handle_notifications.sources import Source, IMAP, get_notifications
+from .test_handler import get_base_notification_data, generate_raw_notification
+
+
+SOURCE_1 = {
+    "alias": "example",
+    "account": "me@example.com",
+    "secret": "supersecret",
+    "url": "imap://example.com",
+}
+
+
+class TestSources(TestCase):
+    """Test case for all the related methods in Email Helper."""
+
+    fixtures = ["handle_notifications_job.yaml"]
+
+    logger = Mock()
+    logger.log_info = Mock()
+    logger.log_warning = Mock()
+
+    def setUp(self):
+        """Prepare data for tests."""
+        settings.PLUGINS_CONFIG = {"nautobot_circuit_maintenance": {"notification_sources": [SOURCE_1.copy()]}}
+        # Deleting other NotificationSource to define a reliable state.
+        NotificationSource.objects.exclude(alias=SOURCE_1["alias"]).delete()
+
+    def test_source_factory(self):
+        """Validate Factory pattern for Source class."""
+        source_instance = Source.init(alias=SOURCE_1["alias"])
+        self.assertIsInstance(source_instance, IMAP)
+        self.assertEqual(source_instance.alias, SOURCE_1["alias"])
+        self.assertEqual(source_instance.url, SOURCE_1["url"])
+        self.assertEqual(source_instance.user, SOURCE_1["account"])
+        self.assertEqual(source_instance.password, SOURCE_1["secret"])
+        self.assertEqual(source_instance.imap_server, "example.com")
+        self.assertEqual(source_instance.imap_port, 993)
+
+    def test_source_factory_nonexistent_alias(self):
+        """Validate Factory pattern for non existent alias."""
+        source_instance = Source.init(alias="non existent alias")
+        self.assertEqual(source_instance, None)
```

Comment:
Would it be better for `init` to raise an exception instead of returning None in these cases?

Reply:
Yes, I will implement it with an exception: https://www.kim-lindhard.com/is-it-ok-for-a-factory-to-return-null/